### PR TITLE
Add rule `vue/require-valid-default-prop`.

### LINKED ISSUE #117: Rule Proposal: `vue/require-valid-default-prop`
## Please describe what the rule should do:

Enforces prop default values to be valid. In vue we have 2 ways to define `default` prop value:
- by function
  ```js
  {
    foo: {
      default () {}
    }
  }
  ```
- by literal value
  ```js
  {
    foo: {
      default: ''
    }
  }
  ```
Literal value is not is not valid when type is specified and set to Array or Object

This rule should also check type of literal values and determine if types match.
When variable is passed as default we should omit checking

## What category of rule is this? (place an "X" next to just one item)
`[ ]` Enforces code style
`[x]` Warns about a potential error
`[ ]` Suggests an alternate way of doing something
`[ ]` Other (please specify:)

## Valid:
```js
Vue.component('example', {
  props: {
    // basic type check (`null` means accept any type)
    propA: Number,
    // multiple possible types
    propB: [String, Number],
    // a number with default value
    propD: {
      type: Number,
      default: 100
    },
    // object/array defaults should be returned from a factory function
    propE: {
      type: Object,
      default: function () {
        return { message: 'hello' }
      }
    }
  }
})
```

## Invalid:
```js
Vue.component('example', {
  props: {
    propA: {
      type: String,
      default: {}
    },
    propB: {
      type: String,
      default: []
    },
    propC: {
      type: Object,
      default: []
    },
    propD: {
      type: Array,
      default: []
    },
    propE: {
      type: Object,
      default: { message: 'hello' }
    }
  }
})
```

see more at: https://vuejs.org/v2/guide/components.html#Prop-Validation

### PR DESCRIPTION
**Pull request by armano2**
This PR implements rules proposed in #117

**TODO:**
- [x] Create documtation
- [x] Add tests
- [x] Implement rule
- [x] Add test & implementation for array of types

fixes #117

### CODE REVIEW

File: tests/lib/rules/require-valid-default-prop.js [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+      parserOptions,
+      errors: errorMessage('function')
+    },
+    {
+      filename: 'test.vue',
+      code: `export default {
+        props: {
+          foo: {
+            type: Array,
+            default: []
```
Comment (michalsnik) [inline tests/lib/rules/require-valid-default-prop.js]: Did you think about supporting also:
```
{
  type: Array,
  default() {
    return 100
  }
}
```
? It should also throw an error I think

File: tests/lib/rules/require-valid-default-prop.js [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+      }`,
+      parserOptions
+    },
+    {
+      filename: 'test.vue',
+      code: `Vue.component('example', {
+        props: {
+          foo: null,
+          foo: Number,
+          foo: [String, Number],
```
Comment (michalsnik) [inline tests/lib/rules/require-valid-default-prop.js]: What if I use it like this:
```
foo: {
  type: [Object, Number],
  default: 10
},
bar: {
  type: [Object, Number],
  default: {}
}
```
? Will it throw an error?

Comment (armano2) [inline tests/lib/rules/require-valid-default-prop.js]: `foo` should not throw an error (it's working)
`bar` should throw an error (it's not working)

File: tests/lib/rules/require-valid-default-prop.js (hunk 1 above)
Comment (armano2) [inline tests/lib/rules/require-valid-default-prop.js]: for now i'm not checking what function is returning or even if function is returning something

File: tests/lib/rules/require-valid-default-prop.js (hunk 2 above)
Comment (armano2) [inline tests/lib/rules/require-valid-default-prop.js]: I'm missing tests for that, i will add them

**Comment by armano2** [comment]
@michalsnik i added few more tests and i applied you suggestions about doc/message

but, we should not check what is inside default function.
its producing to many cases, but if you want me to i can try to handle them.

**Pull request merged**